ДЗ-5

metallheart opened the pull request.

Commit:
ДЗ-5

Commit:
ДЗ-5-fix-tasks-list

bravik commented on app/Services/Tasks/Commands/CommandDTO.php:
```diff
@@ -0,0 +1,19 @@
+
+use Carbon\Carbon;
+
+final readonly class CommandDTO
```
нет смысла в постфиксе DTO

bravik replied:
👍

bravik commented on app/Services/Tasks/Commands/CommandDTO.php:
```diff
@@ -0,0 +1,19 @@
+        public string $title,
+        public string $description,
+        public int $executor_id,
+        public int $category_id,
```
именование полей не по PSR, snake_case не рекоммендуется

bravik replied:
👍

bravik commented on app/Services/Tasks/Handlers/CreateHandler.php:
```diff
@@ -0,0 +1,28 @@
+use App\Repositories\Tasks\TaskRepositoryInterface;
+use App\Services\Tasks\Commands\CommandDTO;
+
+class CreateHandler{
```
Я бы порекомендовал группировать команды и их обработчики вместе по бизнес функции / фиче, а не по конструкции вроде Command/Handler , но в целом вы же на вебинаре видели как я делал, значит делаете это намеренно... Выбор ваш

bravik replied:
👍

bravik commented on app/Repositories/Tasks/TaskRepository.php:
```diff
@@ -0,0 +1,62 @@
+    /**
+     * @param Task $task
+     */
+    public function __construct(private Task $task)
```
Как и зачем у вас в конструкторе оказалась модель? Репозиторий-сервис, он не должен принимать модель в конструктор, только другие сервисы-зависимости в случае необходимости

bravik replied:
👍

bravik commented on app/Repositories/Tasks/TaskRepository.php:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    /**
+     * @return array
```
должен быть указан тип...
array<array-key,Task> или Task[]

bravik commented on app/Repositories/Tasks/TaskRepository.php:
```diff
@@ -0,0 +1,62 @@
+     * @param int $perPage
+     * @return LengthAwarePaginator
+     */
+    public function getAllPaginated(int $perPage = 10): LengthAwarePaginator {
```
Тут надо иметь в виду, что вы из репозитория возвращаете LengthAwarePaginator, который часть пакета Eloquent. Если все хотите сделать чисто в плане архитектуры, то репозиторий не должен возвращать никаких посторонних объектов, только свои собственные. То есть тут нужно делать свой пагинатор и адаптер.

Сейчас можно это проигнорировать, но понимать нужно

bravik replied:
> Так же могли бы показать как правильно реализовать пагинацию?
У вас сейчас LengthAwarePaginator прокидывается насквозь через весь код с уровня инфраструктуры из репозитория, потом на уровень бизнес логики в сервис (fetcher), потом на уровень презентации в контроллеры, потом в шаблон. И все это вы делаете судя по всему только чтобы воспользоваться ларавелевской фичей $paginator->links() которая ссылки отрендерит. 

кстати под капотом когда вы дерагете метод ->paginate() в Eloquent - LengthAwarePaginator лезет через сервис локатор в контейнер и вытасиквает Request, что как бы опять нарушает границы слове.. запрос - слой презентации, работа с базой - инфраструктуры.

Если вы стремитесь четко разграничить уровни, то вы нужно все делать явно:
1) На уровне бизнес логике в фетчере сделать свой класс PaginatedResult. Это очень простой класс который содержит а) подмножество вытаскиваемых сущностей 2) счетчик сколько всего сущностей в базе 3) offset/limit - сколько в странице элементов и какая страница или какое смещение относительно начала.
2) В репозитории вы можете сделать явный метод `fetchPaginated(int $limit, int $offset): Task[]` или `fetchPaginated(int $page, int $pageSize): Task[]`. 
3) В фетчере вы дергаете метод репозитория и передаете туда limit и offset (или page, pageSize) из Query и из результатов создаете свой объект Paginator.
4) Далее вы же хотите ссылки чтобы сами сгенерировались  шаблоне... поэтому вы можете прямо в контроллере PaginatedResults прежде чем отправить в шаблон, преобразовать в LengthAwarePaginator. Просто передать ему все в конструктор и в шаблоне у вас все останется по прежнему.

bravik commented on app/Repositories/Tasks/TaskRepository.php:
```diff
@@ -0,0 +1,62 @@
+
+    /**
+     * @param int $id
+     * @return Task|null
```
Уже есть нативные типы, смысл дублироать

bravik commented on app/Repositories/Tasks/TaskRepository.php:
```diff
@@ -0,0 +1,62 @@
+    /**
+     * @return Task
+     */
+    public function create(): Task {
```
Репозиторий не должен создавать модели, это не его отетственность. Вы это делаете либо в своих сервисах либо в фабриках. Репозиторий работает с уже созданными объектами, добавляет в коллекцию (БД) или ищет...

bravik replied:
👍

bravik commented on app/Services/Categories/Handlers/CreateHandler.php:
```diff
@@ -0,0 +1,22 @@
+
+    public function __invoke(CommandDTO $commandDTO): bool
+    {
+        $category = $this->categoryRepository->create();
```
В след за комменатрием к репозиторию, тут вы делегируете создание объекта репозиторию. А он этим заниматься не должен. Тут создавайте прямо модель.

bravik replied:
👍

bravik commented on app/Services/Categories/Handlers/DestroyHandler.php:
```diff
@@ -0,0 +1,23 @@
+<?php
+namespace App\Services\Categories\Handlers;
+
+use App\Services\Categories\Exceptions\CategoryNotFoundException;
```
👍🏻

bravik commented on app/Services/Categories/Handlers/EditHandler.php:
```diff
@@ -0,0 +1,38 @@
+use App\Services\Categories\Results\CategoryDTO;
+use App\Repositories\Categories\CategoryRepositoryInterface;
+
+class EditHandler
```
Именование хэндлер создает ожидания, что будет команда, а тут запрос а не команда.
Для обработчиков запросов уместнее именование Fetcher. + следует разделять обработчики запросов и обработчики команд следуя принципу CQRS

У вас я вижу один фетчер есть, но он почему-то в папке Results...

Я бы рекомендовал такую структуру:
Commands
-- CreateUser
-- -- Command
-- -- Handler
Queries
-- FetchAllTasks
-- -- Query
-- -- Fetcher

bravik replied:
👍

bravik commented on app/Services/Tasks/Results/Fetcher.php:
```diff
@@ -0,0 +1,46 @@
+<?php
+namespace App\Services\Tasks\Results;
```
непонятный нейминг... почему Results?

bravik replied:
👍

bravik commented on app/Services/Tasks/Results/TaskDTO.php:
```diff
@@ -0,0 +1,45 @@
+        ?string $created_at = null,
+        ?string $updated_at = null,
+    ) {
+        // Преобразуем строки дат в объекты Carbon
```
Вот это неплохо бы чтобы в модели было, чтобы уже возвращалась дата.
Для этого есть casts и можно в нужный тип преобразовать, либо вручную в геттере getCreatedAt() создать объект DateTime

bravik replied:
👍

bravik commented on app/Http/Controllers/Admin/Categories/CreateController.php:
```diff
@@ -0,0 +1,33 @@
+    {
+        $request->validated();
+
+        $category = $handler(new CommandDTO($request->get('name'), $request->get('color'), $request->get('description')));
```
А что если возникнет ошибка? Категория с таким же именем уже сущестует например

Commit:
ДЗ-5-env-db-conf

Commit:
ДЗ-5-рефакторинг

#Переименовал CommandDTO в Command
#Именование полей по PSR
#Репозиторий не принимает модель в конструктор
#Убрал создание моделей из репозиториев
#Убрал преобразование дат из DTO
#Рефакторинг структуры сервисов

Commit:
ДЗ-6-миграция-для-поля-статус

bravik commented on app/Http/Controllers/Admin/Categories/CreateController.php:
```diff
@@ -0,0 +1,39 @@
+            description: $request->get('description')
+        );
+
+        $handler->handle($command);
```
Нет обработки ошибок

bravik commented on app/Http/Controllers/Admin/Tasks/CreateController.php:
```diff
@@ -0,0 +1,49 @@
+     */
+    public function create()
+    {
+        $users = User::all();
```
Вытаскиваете в обход query/fetcher и даже без репозиториев

bravik commented on app/Http/Requests/CreateCategoryRequest.php:
```diff
@@ -0,0 +1,31 @@
+    public function rules()
+    {
+        return [
+            'name' => ['required', 'string', 'max:255', 'unique:categories'],
```
Тут unique валидатор полезт в базу сам через eloquent в обход вашей абстрации над базой в виде репозиториев. Править не надо, но понимать надо

bravik commented on app/Services/Commands/CreateUser/Handler.php:
```diff
@@ -0,0 +1,29 @@
+
+use App\Models\User;
+use App\Repositories\Users\UserRepositoryInterface;
+use Illuminate\Support\Facades\Hash;
```
Фасад прямо в сервисах.. Если стремитесь к чистой бизнес логике, такого быть не должно. Можнос делать свой PasswordHasherInterface и на слое инфраструктуры сделать простейшую реализацию которая может хоть этот фасад дергать.. Но не в бизнес логике напрямую

Commit:
ДЗ-5-пагинация.

Commit:
ДЗ-5-обработка ошибок

Commit:
ДЗ-5-query/fetcher

Commit:
ДЗ-5-проверка на уникальность перенесена в репозитории

Commit:
ДЗ-5-убран фасад из сервиса
